File: Core Data/Core Data Sample Codes/iPhoneCoreDataRecipes/iPhoneCoreDataRecipes/Classes/RecipesAppDelegate.m
/*
 Copyright (C) 2017 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 Application delegate that sets up a tab bar controller with two view controllers -- a navigation controller that in turn loads a table view controller to manage a list of recipes, and a unit converter view controller.
 */

#import "RecipesAppDelegate.h"
#import "RecipeListTableViewController.h"
#import "Recipe.h"

@interface RecipesAppDelegate ()

@property (nonatomic, strong) NSManagedObjectModel *managedObjectModel;
@property (nonatomic, strong) NSManagedObjectContext *managedObjectContext;
@property (nonatomic, strong) NSPersistentStoreCoordinator *persistentStoreCoordinator;

@end


#pragma mark -

@implementation RecipesAppDelegate

// The app delegate must implement the window @property
// from UIApplicationDelegate @protocol to use a main storyboard file.
//
@synthesize window;

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    
    // pass down our managedObjectContext to our RecipeListTableViewController
    UITabBarController *tabBarController = (UITabBarController *)self.window.rootViewController;
    UINavigationController *navController = tabBarController.viewControllers[0];
    
    RecipeListTableViewController *recipeListVC = (RecipeListTableViewController *)navController.topViewController;
    recipeListVC.managedObjectContext = self.managedObjectContext;
    
    return YES;
}

// Saves changes in the application's managed object context before the application terminates.
//
- (void)applicationWillTerminate:(UIApplication *)application {
	
    NSError *error;
    if (self.managedObjectContext != nil) {
        if (self.managedObjectContext.hasChanges && ![self.managedObjectContext save:&error]) {
			/*
			 Replace this implementation with code to handle the error appropriately.
			 
			 abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development. If it is not possible to recover from the error, display an alert panel that instructs the user to quit the application by pressing the Home button.
			 */
			NSLog(@"Unresolved error %@, %@", error, error.userInfo);
			abort();
        } 
    }
}


#pragma mark - Core Data stack

/**
 Returns the managed object context for the application.
 If the context doesn't already exist, it is created and bound to the persistent store coordinator for the application.
 */
- (NSManagedObjectContext *)managedObjectContext {
	
    if (_managedObjectContext != nil) {
        return _managedObjectContext;
    }
	
    NSPersistentStoreCoordinator *coordinator = self.persistentStoreCoordinator;
    if (coordinator != nil) {
        _managedObjectContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSMainQueueConcurrencyType];
        _managedObjectContext.persistentStoreCoordinator = coordinator;
    }
    return _managedObjectContext;
}

/**
 Returns the managed object model for the application.
 If the model doesn't already exist, it is created by merging all of the models found in the application bundle.
 */
- (NSManagedObjectModel *)managedObjectModel {
	
    if (_managedObjectModel != nil) {
        return _managedObjectModel;
    }
    _managedObjectModel = [NSManagedObjectModel mergedModelFromBundles:nil];
    return _managedObjectModel;
}

/**
 Returns the URL to the application's documents directory.
 */
- (NSURL *)applicationDocumentsDirectory
{
    return [[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask].lastObject;
}

/**
 Returns the persistent store coordinator for the application.
 If the coordinator doesn't already exist, it is created and the application's store added to it.
 */
- (NSPersistentStoreCoordinator *)persistentStoreCoordinator {
	
    if (_persistentStoreCoordinator != nil) {
        return _persistentStoreCoordinator;
    }
		
	// Copy the default store (with a pre-populated data) into our Documents folder.
    //
    NSString *documentsStorePath =
        [[self applicationDocumentsDirectory].path stringByAppendingPathComponent:@"Recipes.sqlite"];
	
    // if the expected store doesn't exist, copy the default store
	if (![[NSFileManager defaultManager] fileExistsAtPath:documentsStorePath]) {
		NSString *defaultStorePath = [[NSBundle mainBundle] pathForResource:@"Recipes" ofType:@"sqlite"];
		if (defaultStorePath) {
			[[NSFileManager defaultManager] copyItemAtPath:defaultStorePath toPath:documentsStorePath error:NULL];
		}
	}

    _persistentStoreCoordinator =
        [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:self.managedObjectModel];
    
    // Add the default store to our coordinator.
    NSError *error;
    NSURL *defaultStoreURL = [NSURL fileURLWithPath:documentsStorePath];
   if (![_persistentStoreCoordinator addPersistentStoreWithType:NSSQLiteStoreType
                                                  configuration:nil
                                                            URL:defaultStoreURL
                                                        options:nil
                                                          error:&error]) {
		/*
		 Replace this implementation with code to handle the error appropriately.
		 
		 abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development. If it is not possible to recover from the error, display an alert panel that instructs the user to quit the application by pressing the Home button.
		 
		 Typical reasons for an error here include:
		 * The persistent store is not accessible
		 * The schema for the persistent store is incompatible with current managed object model
		 Check the error message to determine what the actual problem was.
		 */
		NSLog(@"Unresolved error %@, %@", error, error.userInfo);
		abort();
    }
    
    // setup and add the user's store to our coordinator
    NSURL *userStoreURL = [[self applicationDocumentsDirectory] URLByAppendingPathComponent:@"UserRecipes.sqlite"];
    if (![_persistentStoreCoordinator addPersistentStoreWithType:NSSQLiteStoreType
                                                                     configuration:nil
                                                                               URL:userStoreURL
                                                                           options:nil
                                                                             error:&error]) {
		/*
		 Replace this implementation with code to handle the error appropriately.
		 
		 abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development. If it is not possible to recover from the error, display an alert panel that instructs the user to quit the application by pressing the Home button.
		 
		 Typical reasons for an error here include:
		 * The persistent store is not accessible
		 * The schema for the persistent store is incompatible with current managed object model
		 Check the error message to determine what the actual problem was.
		 */
		NSLog(@"Unresolved error %@, %@", error, error.userInfo);
		abort();
    }
		
    return _persistentStoreCoordinator;
}

@end

File: Core Data/Core Data Sample Codes/iPhoneCoreDataRecipes/Recipes_Swift/Recipes_Swift/RecipesAppDelegate.swift
/*
Copyright (C) 2017 Apple Inc. All Rights Reserved.
See LICENSE.txt for this sample’s licensing information

Abstract:
Application delegate that sets up a tab bar controller with two view controllers -- a navigation controller that in turn loads a table view controller to manage a list of recipes, and a unit converter view controller.
*/

import UIKit
import CoreData

@UIApplicationMain
class RecipesAppDelegate: UIResponder, UIApplicationDelegate {

	var window: UIWindow?

	
	func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
		// Override point for customization after application launch.
		
		if let tabbBarController = self.window?.rootViewController as? UITabBarController {
			if let navController = tabbBarController.viewControllers?.first as? UINavigationController {
				if let recipeListTVC = navController.topViewController as? RecipeListTableViewController {
					recipeListTVC.managedObjectContext = self.managedObjectContext
				}
			}
		}
		
		return true
	}


	func applicationWillTerminate(_ application: UIApplication) {
		// Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
		
		if self.managedObjectContext.hasChanges {
			try! self.managedObjectContext.save()
		}
		
	}
	
	
	//MARK: - Core Data stack
	
	/**
	Returns the managed object context for the application.
	If the context doesn't already exist, it is created and bound to the persistent store coordinator for the application.
	*/
	lazy var managedObjectContext: NSManagedObjectContext = {
		let coordinatior = self.persistentStoreCoordinator
		let moc = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
		moc.persistentStoreCoordinator = coordinatior
		return moc
	}()
	
	/**
	Returns the managed object model for the application.
	If the model doesn't already exist, it is created by merging all of the models found in the application bundle.
	*/
	lazy var managedObjectModel: NSManagedObjectModel = {
		let modelURL = Bundle.main.url(forResource: "Recipes_Swift", withExtension: "momd")
		let mom = NSManagedObjectModel(contentsOf: modelURL!)
		return mom!
	}()
	
	/**
	Returns the URL to the application's documents directory.
	*/
	var applicationDocumentsDirectory: URL {
		return FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
	}
	
	/**
	Returns the persistent store coordinator for the application.
	If the coordinator doesn't already exist, it is created and the application's store added to it.
	*/
	lazy var persistentStoreCoordinator: NSPersistentStoreCoordinator = {
		// Copy the default store (with a pre-populated data) into our Documents folder.
		//
		let documentsStoreURL = self.applicationDocumentsDirectory.appendingPathComponent("Recipes.sqlite")
		
		// if the expected store doesn't exist, copy the default store
		if !FileManager.default.fileExists(atPath: documentsStoreURL.path) {
			if let defaultStoreURL = Bundle.main.url(forResource: "Recipes", withExtension: "sqlite") {
				try! FileManager.default.copyItem(at: defaultStoreURL, to: documentsStoreURL)
			}
		}
		
		let coordinator = NSPersistentStoreCoordinator(managedObjectModel: self.managedObjectModel)
		
		// Add the default store to our coordinator.
		
		do {
			try coordinator.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: documentsStoreURL, options:  [NSMigratePersistentStoresAutomaticallyOption: true,
																																	NSInferMappingModelAutomaticallyOption: true])
			
//			// setup and add the user's store to our coordinator
//			let userStoreURL = self.applicationDocumentsDirectory.appendingPathComponent("UserRecipes.sqlite")
//			print(userStoreURL)
//			print(documentsStoreURL)
//			try coordinator.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: userStoreURL, options: [NSMigratePersistentStoresAutomaticallyOption: true,
//																															  NSInferMappingModelAutomaticallyOption: true])

		}
		catch {
			print("persistentStore Coordinator Error", error)
		}
		
		return coordinator
		
	}()


}

